courier-imap-4.9.2/maildirmake.1
'\" t
.\"<!-- Copyright 1998 - 2009 Double Precision, Inc.  See COPYING for -->
.\"<!-- distribution information. -->
.\"     Title: maildirmake
.\"    Author: Sam Varshavchik
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: 04/04/2011
.\"    Manual: Double Precision, Inc.
.\"    Source: Courier Mail Server
.\"  Language: English
.\"
.TH "MAILDIRMAKE" "1" "04/04/2011" "Courier Mail Server" "Double Precision, Inc."
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
maildirmake \- create maildirs and maildir folders
.SH "SYNOPSIS"
.HP \w'\fBmaildirmake\fR\ 'u
\fBmaildirmake\fR [options...] {\fImaildir\fR}
.SH "DESCRIPTION"
.PP
The
\fBmaildirmake\fR
command creates maildirs, and maildir folders\&. This documentation describes the
\fBmaildirmake\fR
command from the
Courier
mail server, which creates an extended form of maildirs that implements additional extensions beyond the basic maildir properties that were first implemented in the Qmail mail server\&.
.SH "OPTIONS"
.PP
\-S
.RS 4
create a "sharable" maildir\&. A sharable maildir has slightly different permissions which allows creation of publicly\-shared folders\&.
.RE
.PP
\-q \fIquota\fR
.RS 4
install a quota on the maildir\&. See
\m[blue]\fB\fBmaildirquota\fR(7)\fR\m[]\&\s-2\u[1]\d\s+2, below\&.
.RE
.PP
\-f \fIfolder\fR
.RS 4
do not create a maildir, but create a folder in an existing maildir\&.
.RE
.PP
\-s \fImode\fR
.RS 4
create a publicly accessible folder in an existing sharable maildir\&. First, use the
\fB\-S\fR
option to create a sharable maildir\&. Then, run
\fBmaildirmake\fR
again with the
\fB\-s\fR
option to create publicly accessible folders\&.
\fImode\fR
is a comma\-separated list of the following keywords:
read
\- readonly folder, only you can write messages to this folder;
write
\- anyone can read and write messages to this folder;
group
\- only allow members of your own system group to access messages in this folder (instead of everyone)\&.
.RE
.PP
\-\-add \fIname\fR=\fIpathname\fR, \-\-del \fIname\fR
.RS 4
create or delete the directories and links needed to access shared folders\&. See below for more information\&.
.RE
.SS "FOLDERS"
.PP
This
\fBmaildirmake\fR
command supports enhanced maildirs that contain folders\&.
.PP
By itself,
\fBmaildirmake\fR
makes a new subdirectory
\fImaildir\fR, and creates all the necessary structures\&. The
\fB\-f\fR
option creates a new "folder" within an existing
\fImaildir\fR\&.
\fImaildir\fR
must already exist, and the
\fBmaildirmake\fR
command will create a new folder in the maildir\&.
.PP
Folders are simply subdirectories inside the main maildir whose names start with a period, and which are themselves maildirs\&. For example, the command "\fBmaildirmake \-f Drafts mail/Maildir\fR" creates
mail/Maildir/\&.Drafts, that has the usual
tmp,
new
and
cur\&. You MUST use the
\fB\-f\fR
option, instead of specifying
mail/Maildir/\&.Drafts
directly, in order to correctly initialize certain enhanced maildir features\&.
.PP
Folders cannot be created directly within other folders\&. Running
\fBmaildirmake \-f Urgent mail/Maildir/\&.Drafts\fR
will not work\&. Instead, the period character is designated as a hierarchy separator, run
\fBmaildirmake \-f Drafts\&.Urgent mail/Maildir\fR
instead\&. This creates
mail/Maildir/\&.Drafts\&.Urgent, and all mail software that supports enhanced maildirs will interpret it as a subfolder Urgent of the Drafts folder\&.
.SS "SHARED FOLDERS"
.PP
This is another extension to the Maildir format that allows folders to be shared between multiple clients\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.PP
The
Courier
IMAP server implements two types of shared folders: filesystem permission\-based shared folders, as well as virtual shared folders based on IMAP access control lists\&. Use the
\fBmaildirmake\fR
command to implement shared folders based on filesystem permissions\&. The
\m[blue]\fB\fBmaildiracl\fR(1)\fR\m[]\&\s-2\u[2]\d\s+2
command manages access control lists, which are used by virtual shared folders\&.
.PP
See the
Courier
IMAP server documentation for more information\&.
.sp .5v
.RE
.PP
First, you need to create a collection of sharable folders, as a separate maildir:
.sp
.if n \{\
.RS 4
.\}
.nf
\fBmaildirmake \-S /usr/local/share/maildirs/notices\fR
.fi
.if n \{\
.RE
.\}
.PP
Then, create individuals folders that will be accessed in shared mode:
.sp
.if n \{\
.RS 4
.\}
.nf
\fBmaildirmake \-s write \-f Weekly /usr/local/share/maildirs/notices\fR
.fi
.if n \{\
.RE
.\}
.PP
In this example, the "Weekly" folder is created, with read/write access to everyone\&. Multiple folders can be created in the same maildir, with different access permissions\&. Everyone can create a sharable maildir\&. The access privileges for individual folders are set by the
\fB\-s\fR
option, and are implemented using traditional filesystem permissions\&.
.PP
Use the
\fB\-\-add\fR
and
\fB\-\-del\fR
options to add a sharable maildir to an existing maildir\&. Client software that implements this extension will now know where to find sharable folders:
.sp
.if n \{\
.RS 4
.\}
.nf
\fBmaildirmake \-\-add notices=/usr/local/share/maildirs/notices $HOME/Maildir\fR
.fi
.if n \{\
.RE
.\}
.PP

$HOME/Maildir
is your main maildir\&. The argument to
\fB\-add\fR
is
\fInick\fR=\fIpath\fR\&.
\fInick\fR
is a nickname for this collection of sharable folders, and
\fIpath\fR
is the location of the sharable maildir\&. All folders in the sharable maildir that you have access to \-\- such as "Weekly", in this case, will now be accessible\&. Multiple sharable maildirs can be added, by giving each one a unique
\fInick\fR\&.
.PP
The
\fB\-\-del\fR
option "disconnects" the sharable maildir from the main maildir\&.
.SS "GLOBAL SHARED FOLDERS"
.PP
Normally
\fB\-add\fR
command must be run for every maildir which needs to access the sharable maildir\&. Alternatively the file
/var/indimail/etc/maildirshared
can be created, to specify a default set of sharable maildirs\&. Each line in this file takes the following format:
.sp
.if n \{\
.RS 4
.\}
.nf
\fInick\fR<tab>\fIpath\fR
.fi
.if n \{\
.RE
.\}
.PP
\fInick\fR
is a short nickname for the sharable maildir,
<tab>
is a single tab character,
\fIpath\fR
is the pathname to the sharable maildir\&.
.SS "ACCESSING SHARED FOLDERS"
.PP
You may have read or write access to a shared folder\&. If you have write access, you can add messages to the shared folder\&. You can also delete messages that you\'ve added\&.
.PP
Anyone can create a sharable maildir, so if the sharable maildir is actually created by you, can can delete any message, not just your own\&.
.SH "SEE ALSO"
.PP

\m[blue]\fB\fBmaildir\fR(5)\fR\m[]\&\s-2\u[3]\d\s+2,
\m[blue]\fB\fBmaildiracl\fR(1)\fR\m[]\&\s-2\u[2]\d\s+2,
\m[blue]\fB\fBmaildirkw\fR(1)\fR\m[]\&\s-2\u[4]\d\s+2,
\m[blue]\fB\fBmaildrop\fR(1)\fR\m[]\&\s-2\u[5]\d\s+2,
\m[blue]\fB\fBmaildirquota\fR(7)\fR\m[]\&\s-2\u[1]\d\s+2,
\m[blue]\fB\fBdeliverquota\fR(8)\fR\m[]\&\s-2\u[6]\d\s+2,
\m[blue]\fB\fBmaildropfilter\fR(7)\fR\m[]\&\s-2\u[7]\d\s+2,
\m[blue]\fBhttp://www\&.qmail\&.org/man/man5/maildir\&.html\fR\m[]\&.
.SH "AUTHOR"
.PP
\fBSam Varshavchik\fR
.RS 4
Author
.RE
.SH "NOTES"
.IP " 1." 4
\fBmaildirquota\fR(7)
.RS 4
\%[set $man.base.url.for.relative.links]/maildirquota.html
.RE
.IP " 2." 4
\fBmaildiracl\fR(1)
.RS 4
\%[set $man.base.url.for.relative.links]/maildiracl.html
.RE
.IP " 3." 4
\fBmaildir\fR(5)
.RS 4
\%[set $man.base.url.for.relative.links]/maildir.html
.RE
.IP " 4." 4
\fBmaildirkw\fR(1)
.RS 4
\%[set $man.base.url.for.relative.links]/maildirkw.html
.RE
.IP " 5." 4
\fBmaildrop\fR(1)
.RS 4
\%[set $man.base.url.for.relative.links]/maildrop.html
.RE
.IP " 6." 4
\fBdeliverquota\fR(8)
.RS 4
\%[set $man.base.url.for.relative.links]/deliverquota.html
.RE
.IP " 7." 4
\fBmaildropfilter\fR(7)
.RS 4
\%[set $man.base.url.for.relative.links]/maildropfilter.html
.RE


courier-imap-4.9.0/maildir/maildirmake.c
/*
** Copyright 1998 - 2006Double Precision, Inc.
** See COPYING for distribution information.
*/


#if	HAVE_CONFIG_H
#include	"config.h"
#endif
#include	<stdio.h>
#include	<string.h>
#include	<stdlib.h>
#if	HAVE_FCNTL_H
#include	<fcntl.h>
#endif
#if	HAVE_UNISTD_H
#include	<unistd.h>
#endif
#include	<sys/types.h>
#if	HAVE_SYS_STAT_H
#include	<sys/stat.h>
#endif
#include	<ctype.h>
#include	<errno.h>

#include	"maildircreate.h"
#include	"maildirmisc.h"
#include	"maildirsharedrc.h"
#include	"maildirquota.h"

static void usage()
{
	printf("Usage: maildirmake [ options ] maildir\n");
	exit(1);
}

extern FILE *maildir_shared_fopen(const char *, const char *);
extern void maildir_shared_fparse(char *, char **, char **);

static void add(const char *dir, const char *name)
{
char	*c=malloc(strlen(name)+1);
char	*s;
FILE	*fp;
char	buf[BUFSIZ];
char	*ptr, *dirp;

	if (!c)
	{
		perror("malloc");
		exit(1);
	}
	strcpy(c, name);

	if ((s=strchr(c, '=')) == 0)
		usage();
	*s++=0;
	if (*s != '/')
		usage();
	if (access(s, R_OK))
	{
		perror(s);
		exit(1);
	}
	if (strchr(c, '.') || strchr(c, '/'))
	{
		fprintf(stderr, "%s: invalid name\n", c);
		exit(1);
	}

	if (chdir(dir))
	{
		perror(dir);
		exit(1);
	}

	if ((fp=fopen(MAILDIRSHAREDRC, "r")) != 0)
	{
		while ((ptr=fgets(buf, sizeof(buf), fp)) != 0)
		{
		char	*namep;

			maildir_shared_fparse(buf, &namep, &dirp);
			if (!namep)	continue;
			if (strcmp(namep, c) == 0)
			{
				fclose(fp);
				fprintf(stderr,
		"%s: already defined as a sharable maildir in %s.\n",
					namep, MAILDIRSHAREDRC);
				exit(2);
			}
		}
		fclose(fp);
	}

	if ((fp=maildir_shared_fopen(".", "a+")) == 0)
	{
		perror(dir);
		exit(1);
	}

	if (fseek(fp, 0L, SEEK_SET) < 0)
	{
		perror(dir);
		exit(1);
	}

	while ((ptr=fgets(buf, sizeof(buf), fp)) != 0)
	{
	char	*namep;

		maildir_shared_fparse(buf, &namep, &dirp);
		if (!namep)	continue;
		if (strcmp(namep, c) == 0)
		{
			fclose(fp);
			fprintf(stderr, "%s: already defined as a sharable maildir.\n",
				namep);
			exit(2);
		}
	}
	fprintf(fp, "%s\t%s\n", c, s);
	if (fflush(fp) || ferror(fp) || fclose(fp))
	{
		perror("dir");
		exit(1);
	}
	exit(0);
}

static void del(const char *dir, const char *n)
{
FILE	*fp;
char	buf[BUFSIZ];
char	buf2[BUFSIZ];

FILE	*fp2;
int	found;
 struct maildir_tmpcreate_info createInfo;

	if (chdir(dir))
	{
		perror(dir);
		exit(1);
	}
	if ((fp=maildir_shared_fopen(".", "r")) == 0)
	{
		perror(dir);
		exit(1);
	}

	maildir_tmpcreate_init(&createInfo);

	createInfo.uniq="shared";
	createInfo.msgsize=0;
	createInfo.doordie=1;

	if ((fp2=maildir_tmpcreate_fp(&createInfo)) == NULL)
	{
		perror(dir);
		exit(1);
	}

	found=0;
	while (fgets(buf, sizeof(buf), fp))
	{
	char	*name, *dirp;

		strcpy(buf2, buf);
		maildir_shared_fparse(buf2, &name, &dirp);
		if (name && strcmp(name, n) == 0)
		{
			found=1;
			continue;
		}
		fprintf(fp2, "%s", buf);
	}
	fclose(fp);
	if (fflush(fp2) || ferror(fp2) || fclose(fp2)
	    || rename(createInfo.tmpname, "shared-maildirs"))
	{
		perror(createInfo.tmpname);
		unlink(createInfo.tmpname);
		exit(1);
	}
	maildir_tmpcreate_free(&createInfo);
	if (!found)
	{
		fprintf(stderr, "%s: not found.\n", n);
		exit(1);
	}
	exit(0);
}

int main(int argc, char *argv[])
{
const char *maildir, *folder=0;
char *p;
int	argn;
int	perm=0700;
int	musthavefolder=0;
int	subdirperm;
char	*addshared=0, *delshared=0;
const char *quota=0;
char	*tmp=0;

	for (argn=1; argn<argc; argn++)
	{
		if (argv[argn][0] != '-')	break;
		if (strcmp(argv[argn], "-") == 0)	break;
		if (strncmp(argv[argn], "-f", 2) == 0)
		{
			folder=argv[argn]+2;
			if (*folder == 0 && argn+1 < argc)
				folder=argv[++argn];
			continue;
		}
		if (strcmp(argv[argn], "-S") == 0)
		{
			perm=0755;
			continue;
		}

		if (strncmp(argv[argn], "-s", 2) == 0)
		{
		const char *p=argv[argn]+2;

			if (*p == 0 && argn+1 < argc)
				p=argv[++argn];

			perm=0755;
			for (; *p; ++p)
			{
				if (isspace((int)(unsigned char)*p) ||
					*p == ',')
					continue;
				if (*p == 'r')
					perm=0755;
				else if (*p == 'w')
					perm=0777 | S_ISVTX;
				else if (*p == 'g')
					perm &= ~0007;

				while (*p && !isspace((int)(unsigned char)*p)
					&& *p != ',')
					++p;
				if (!*p)	break;
			}
			musthavefolder=1;
			continue;
		}

		if (strncmp(argv[argn], "-q", 2) == 0)
		{
			const char *p=argv[argn]+2;

			if (*p == 0 && argn+1 < argc)
				p=argv[++argn];

			quota=p;
			continue;
		}

		if (strcmp(argv[argn], "--add") == 0 && argc-argn > 1)
		{
			addshared=argv[++argn];
			continue;
		}

		if (strcmp(argv[argn], "--del") == 0 && argc-argn > 1)
		{
			delshared=argv[++argn];
			continue;
		}

		usage();
	}

	if (argn == argc)	usage();
	maildir=argv[argn];

	if (addshared)
	{
		add(maildir, addshared);
		exit (0);
	}

	if (delshared)
	{
		del(maildir, delshared);
		exit (0);
	}

	if (folder && *folder == '.')
	{
		printf("Invalid folder name: %s\n", folder);
		exit(1);
	}

	if (folder)
	{
		if ((p=(char *)malloc(strlen(maildir)+strlen(folder)+4)) == 0)
		{
			perror("maildirmake malloc");
			exit(1);
		}
		strcat(strcat(strcpy(p, maildir), "/."), folder);
		maildir=p;
	}
	else	if (musthavefolder)
		usage();

	if (quota)
	{
		maildir_quota_set(maildir, quota);
		exit(0);
	}
	subdirperm=perm;
	if (!folder)	subdirperm=0700;
	umask(0);
	if (maildir_make(maildir, perm & ~0022, subdirperm, folder ? 1:0) < 0)
	{
		tmp=(char *)malloc(strlen(maildir)+14);
		if (!tmp) {
			perror("maildirmake malloc");
			exit(1);
		}
		snprintf(tmp, 14+strlen(maildir), "maildirmake: %s", maildir);
		perror(tmp);
		exit(1);
	}
	exit(0);
	return (0);
}
